core/src/main/kotlin/com/bumble/appyx/core/children/ChildNodeCreationManager.kt

```diff
@@ -0,0 +1,112 @@
+import kotlinx.coroutines.flow.updateAndGet
+import kotlinx.coroutines.launch
+
+internal class ChildNodeCreationManager<Routing : Any>(
+    private var savedStateMap: SavedStateMap?,
+    private val customisations: NodeCustomisationDirectory,
+    private val childMode: ChildEntry.ChildMode,
+) {
+    private val _children =
+        MutableStateFlow<Map<RoutingKey<Routing>, ChildEntry<Routing>>>(emptyMap())
+    val children: StateFlow<ChildEntryMap<Routing>> = _children.asStateFlow()
+
+    fun launch(parentNode: ParentNode<Routing>) {
+        savedStateMap.restoreChildren(parentNode)?.also { restoredMap ->
+            _children.update { restoredMap }
+            savedStateMap = null
+        }
+        parentNode.lifecycle.coroutineScope.launch {
+            parentNode.navModel.elements.collect { elements ->
+                _children.update { map ->
+                    val navModelKeys = elements
+                        .mapTo(HashSet(elements.size, 1f)) { element -> element.key }
+                    val localKeys = map.keys
+                    val newKeys = navModelKeys - localKeys
+                    val removedKeys = localKeys - navModelKeys
+                    val mutableMap = map.toMutableMap()
+                    newKeys.forEach { key ->
+                        mutableMap[key] =
+                            ChildEntry.create(
+                                key = key,
+                                resolver = parentNode,
+                                buildContext = null.toBuildContext(parentNode),
+                                childMode = childMode,
+                            )
+                    }
+                    removedKeys.forEach { key ->
+                        mutableMap.remove(key)
+                    }
+                    mutableMap
+                }
+            }
+        }
+    }
+
+    fun childOrCreate(routingKey: RoutingKey<Routing>): ChildEntry.Eager<Routing> {
+        val value = _children.value
+        val child = value[routingKey] ?: error(
+            "Rendering and children management is out of sync: requested $routingKey but have only ${value.keys}"
+        )
+        return when (child) {
+            is ChildEntry.Eager ->
+                child
+            is ChildEntry.Lazy ->
+                _children.updateAndGet { map ->
+                    val updateChild = map[routingKey]
+                        ?: error("Requested child $routingKey disappeared")
+                    when (updateChild) {
+                        is ChildEntry.Eager -> map
+                        is ChildEntry.Lazy -> map.plus(routingKey to updateChild.initialize())
+                    }
+                }[routingKey] as ChildEntry.Eager
+        }
+    }
+
+    private fun SavedStateMap?.restoreChildren(parentNode: ParentNode<Routing>): ChildEntryMap<Routing>? =
+        (this?.get(KEY_CHILDREN_STATE) as? Map<RoutingKey<Routing>, SavedStateMap>)?.mapValues {
```

Comment:
Worth moving the private functions to the bottom? Also, can they be added to the companion object?

Reply:
It will become non-static in the next PR.

In general Kotlin code style recommends to group methods by context and not visibility, but I do not have any strong opinion on it.